nixos/modules/services/monitoring/grafana.nix

```diff
@@ -50,6 +50,166 @@ let
     SMTP_FROM_ADDRESS = cfg.smtp.fromAddress;
   } // cfg.extraOptions;
 
+  datasourceConfiguration = {
+    apiVersion = 1;
+    datasources = cfg.provision.datasources;
+  };
+
+  datasourceFile = pkgs.writeText "datasource.yaml" (builtins.toJSON datasourceConfiguration);
+
+  dashboardConfiguration = {
+    apiVersion = 1;
+    providers = cfg.provision.dashboards;
+  };
+
+  dashboardFile = pkgs.writeText "dashboard.yaml" (builtins.toJSON dashboardConfiguration);
+
+  provisionConfDir = pkgs.stdenv.mkDerivation {
+    name = "grafana-provisioning";
+    buildCommand = ''
+      mkdir $out
+      mkdir $out/datasources
+      mkdir $out/dashboards
+      ln -sf ${datasourceFile} $out/datasources/datasource.yaml
+      ln -sf ${dashboardFile} $out/dashboards/dashboard.yaml
+    '';
+  };
+
+  # Get a submodule without any embedded metadata:
+  _filter = x: filterAttrs (k: v: k != "_module") x;
+
+  # http://docs.grafana.org/administration/provisioning/#datasources
+  grafanaTypes.datasourceConfig = types.submodule {
+    options = {
+      name = mkOption {
+        type = types.str;
+        default = "Grafana";
+        description = "Name of the datasource. Required";
+      };
+      type = mkOption {
+        type = types.str;
```

Comment:
As there's only a limited amount of possible data sources, we probably want to use something like `types.enum [ "graphite" "prometheus" /* and so on */ ]`